gazebo_plugins/src/gazebo_ros_wheel_slip.cpp

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020 Open Source Robotics Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gazebo/physics/Model.hh>
+#include <gazebo/physics/World.hh>
+
+#include <gazebo/transport/transport.hh>
+#include <gazebo_plugins/gazebo_ros_wheel_slip.hpp>
+#include <gazebo_ros/node.hpp>
+#include <std_msgs/msg/bool.hpp>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace gazebo_plugins
+{
+class GazeboRosWheelSlipPrivate
+{
+public:
+  /// A pointer to the GazeboROS node.
+  gazebo_ros::Node::SharedPtr ros_node_;
+};
+
+GazeboRosWheelSlip::GazeboRosWheelSlip()
+: impl_(std::make_unique<GazeboRosWheelSlipPrivate>())
+{
+}
+
+GazeboRosWheelSlip::~GazeboRosWheelSlip()
+{
+}
+
+void GazeboRosWheelSlip::Load(gazebo::physics::ModelPtr _model, sdf::ElementPtr _sdf)
+{
+  // Initialize the WheelSlipPlugin first so its values are preferred unless the ros
+  // parameters are overridden by a launch file.
+  WheelSlipPlugin::Load(_model, _sdf);
+
+  // Initialize ROS node
+  impl_->ros_node_ = gazebo_ros::Node::Get(_sdf);
+
+  auto param_change_callback =
+    [this](std::vector<rclcpp::Parameter> parameters) {
+      auto result = rcl_interfaces::msg::SetParametersResult();
+      result.successful = true;
+      for (const auto & parameter : parameters) {
+        std::string param_name = parameter.get_name();
+        if (param_name == "slip_compliance_unitless_lateral") {
+          double slip = parameter.as_double();
+          if (slip >= 0.) {
+            RCLCPP_INFO(impl_->ros_node_->get_logger(),
+              "New lateral slip compliance: %.3e", slip);
+            this->SetSlipComplianceLateral(slip);
+          }
+        } else if (param_name == "slip_compliance_unitless_longitudinal") {
+          double slip = parameter.as_double();
+          if (slip >= 0.) {
+            RCLCPP_INFO(impl_->ros_node_->get_logger(),
+              "New longitudinal slip compliance: %.3e", slip);
+            this->SetSlipComplianceLongitudinal(slip);
+          }
+        } else {
+          RCLCPP_WARN(impl_->ros_node_->get_logger(),
+            "Unrecognized parameter name[%s]", param_name);
+          result.successful = false;
+        }
+      }
+      return result;
+    };
+
+  impl_->ros_node_->add_on_set_parameters_callback(param_change_callback);
```

Comment:
You might want to double-check, but I think we should maintain a reference to the returned callback handle, otherwise the callback may be unregistered.

Reply:
yeah, it looks like this was done in https://github.com/ros2/rclcpp/pull/1136

Reply:
I copied the approach from disparity_node in stereo_image_proc: https://github.com/ros-perception/image_pipeline/blob/2.2.0/stereo_image_proc/src/stereo_image_proc/disparity_node.cpp#L92-L182 in 0bef26d